File: src/leetcode/questions/x201/to/x299/Problem_202.java
package leetcode.questions.x201.to.x299;

import java.util.HashSet;
import java.util.Set;

public class Problem_202 {
	private static boolean recursive(int n, Set<Integer> set) {
		if (set.contains(n))return false;
		int output = 0, input =n ;
		while (n > 9) {
			output += Math.pow(n % 10, 2);
			n = n/ 10;
		}
		output += Math.pow(n % 10, 2);
		if (output == 1)return true;
		set.add(input);
		return recursive(output, set);
	}

	public static  boolean isHappy(int n) {
		return recursive(n, new HashSet<Integer>());
	}
}

File: src/leetcode/questions/x201/to/x299/Problem_202.go
package main

import (
	"math"
)

func recursive(n int, set map[int]struct{}) bool {
	if _, exists := set[n]; exists {
		return false
	}
	output := 0
	input := n
	for n > 9 {
		output += int(math.Pow(float64(n%10), 2))
		n = n / 10
	}
	output += int(math.Pow(float64(n%10), 2))
	if output == 1 {
		return true
	}
	set[input] = struct{}{}
	return recursive(output, set)
}

func isHappy(n int) bool {
	return recursive(n, make(map[int]struct{}))
}
